### PR TITLE
New TbotsRobotMsg protos

### LINKED ISSUE #1272: Create TbotsRobotMsg Proto
### Description of the task
Currently, robot statuses such as break beam, capacitor voltage, etc., are sent to the robot in custom encoded packets over radio. The new TbotsRobotMsg Proto will use the Protobuf encoding and standardize this packet for the old and new robots. The new proto design will need to consider the needs of the new robot including new sensors or new sensor data we want to collect and should be done with sufficient consultation.

### Acceptance criteria

- [ ] Develop structure for Tbots_RobotMsg Proto
- [ ] Consult with relevant stakeholders
- [ ] Finalize the proto

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Prototyped a new Tbots_RobotMsg proto structured as `oneof` RobotStatus2014 proto or FrankieStatus. 

### Testing Done
Builds

### Resolved Issues
Resolves #1272

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by jonathanlew** [comment]
This PR is marked WIP because I want finalize the structure of the proto before opening this review up to electrical and other stakeholders. By structure, I would like your thoughts on using oneof and what things should be shared between different versions of the robot - some fields are more hardware-dependent than others.

File: src/shared/proto/BUILD
Comment (garethellis0) [inline src/shared/proto/BUILD]: We should be consistent about how we label the old robots. In the firmware I've labelled them as `legacy_robot_stm32f4`, but I think I prefer the naming you have here.

All that said, why are there different proto's for the old and new robots? Could we have a single proto with optional fields for the fields that the old robots don't supply?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/shared/proto/BUILD
Comment (jonathanlew) [reply src/shared/proto/BUILD]: if we had optional fields for all the fields that are different between the new and old robots, then SensorFusion would have to check before using each one of them, instead of just checking once for the version of the proto.

File: src/shared/proto/tbots_frankie_status.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_frankie_status.proto]: UberNit: Can we remove the mix of snake_case and just use CamelCase for proto

so `Tbots_EncoderData` -> `TbotsEncoderData`

File: src/shared/proto/tbots_frankie_status.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_frankie_status.proto]: We can probably just make this `TbotsRobotStatus`

File: src/shared/proto/tbots_robot_msg.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_robot_msg.proto]: strings are kind of annoying to use with `nanopb` (doable if needed, but annoying so it would be nice if we can avoid)

Is this so robots can send arbitrary debug messages over the network if needed? Or is this used for something else?

File: src/shared/proto/tbots_robot_msg.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_robot_msg.proto]: ditto previous comment, is this just a network log from the robot? What would we use the `repeated string` `network_messages` for over `robot_messages`?

File: src/shared/proto/tbots_robot_status_2014.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_robot_status_2014.proto]: This is probably getting very close to the 100 byte radio limit if the goal is to eventually send this over radio from the old robots. 

We probably need to do some pruning to get it just under 100 bytes + the extra stuff in `TbotsRobotMsg`

```
     // The link quality of the last received packet, from 0 (worst) to 1 (best).
     required double link_quality = 18;

     // The received signal strength of the last received packet, in decibels.
     required int32 received_signal_strength_db = 19;

     // Whether or nott the build ID information is valid.
     required bool build_ids_valid = 20;

     // The microcontroller firmware build ID.
     required uint32 fw_build_id = 21;

     // The FPGA bitstream build ID.
     required uint32 fpga_build_id = 22;
```

We probably don't need these @garethellis0 are we currently using this stuff?

File: src/shared/proto/BUILD
Comment (MathewMacDougall) [reply src/shared/proto/BUILD]: What would the API to the filters be in that case? For example, assuming we move forward with EKFs, they can work with as much or as little data as we have. Would we need two EKFs versions to support each type of proto?

File: src/shared/proto/tbots_frankie_status.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_frankie_status.proto]: Perhaps we should define a style guide for our own Protos now that we are writing our own (will be the same deal for primitive messages)

File: src/shared/proto/tbots_robot_msg.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_robot_msg.proto]: I'm assuming this is legacy stuff that was moved from the old robot status structure. @jonathanlew I'm assuming you essentially copied most of the `RobotStatus` struct from `backend/robot_status.h` when creating these messages.

This old `RobotStatus` struct does contain strings, but they are added as part of the status handling in `backend/output/radio/mrf/annunciator.cpp`, in the `handle_robot_message` function on line `105`. This is where values are checked and if something isn't right, messages from a lookup table are added.

Eg.
```
new_msgs.insert(new_msgs.begin(), MRF::LOW_CAP_MESSAGE);
```

So to my knowledge the robots never sent string over radio, any strings that were added to the `RobotStatus` were from an error code or post-processing, so I don't think we need strings in the proto message.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Holding off on any comments re: fields that we may want to remove. I've replied to a few existing comments. I agree we should:
1. Define a proto style guide
2. Agree on overall message structure (`optional` values, etc.)

before involving ELEC and other stakeholders. At that time I think it would be easiest to just get everyone in a call to go over everything alltogether

File: src/shared/proto/tbots_robot_msg.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_robot_msg.proto]: I see I'll change it to some sort of error code then

File: src/shared/proto/tbots_frankie_status.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_frankie_status.proto]: I tried to imitate the the style of SSL_WrapperPacket naming, but we can define our own style if we'd like

File: src/shared/proto/tbots_frankie_status.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_frankie_status.proto]: If we're going with the oneof structure then the statuses do need to versioned, i.e. 2014, etc. for a hybrid fleet, say

File: src/shared/proto/BUILD
Comment (jonathanlew) [reply src/shared/proto/BUILD]: I'd imagine we'd have functions like filterRobotStatus(Tbots_RobotStatus2014 status) and filterRobotStatus(Tbots_FrankieStatus status)

File: src/shared/proto/tbots_robot_msg.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_robot_msg.proto]: We should maybe sync up about if we want to use predominantly error codes, or determine stuff with the values sent back (like we do currently in the annunciator).

Pros of error codes
* Should work better with a hybrid fleet. The software won't need to check raw values against different thresholds for different hardware
* Can get info / errors about things that aren't sent back explicitly (eg. SD Card errors)

Cons of error codes
* Can't get specific values for the errors. Eg. we can warn about low battery, but we don't know the exact value

IMO we should user error codes as much as possible for reporting faults and errors (primarily used to detect faulty robots for things like substitution, and to report in the Visualizer). All other values should just be for extra info, whether it be debugging or just readability (eg. battery status, cap voltage).

At least this way there's a clear distinction between where errors / faults come from, and where "data" comes from. The software should _not_ perform calculations on raw data to add additional errors.

Thoughts @garethellis0 @jonathanlew @akhilveeraghanta ?

File: src/shared/proto/BUILD
Comment (MathewMacDougall) [reply src/shared/proto/BUILD]: But how would that look internally? Assuming the status messages share some common data (battery, cap voltage, etc), not that these are used in filtering, this seems like it would cause duplication in the filters.

For the sake of argument, what it the only difference between 2 robot generations was that one had an IMU and the other did not?

@garethellis0 for how / if this matters for standard EKF structure.

File: src/shared/proto/tbots_frankie_status.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_frankie_status.proto]: I don't have a strong preference as long as we're consistent. Perhaps easiest to just follow Google's? https://developers.google.com/protocol-buffers/docs/style

Comment (garethellis0) [reply src/shared/proto/tbots_frankie_status.proto]: I would strongly support just following google's, it will prevent debate and make our proto format more similar to others.

File: src/shared/proto/tbots_robot_msg.proto
Comment (garethellis0) [reply src/shared/proto/tbots_robot_msg.proto]: Sounds good.

File: src/shared/proto/tbots_robot_status_2014.proto
Comment (garethellis0) [reply src/shared/proto/tbots_robot_status_2014.proto]: Not to the best of my knowledge, but I defer to the team historian @MathewMacDougall

File: src/shared/proto/BUILD
Comment (garethellis0) [reply src/shared/proto/BUILD]: Hmm. Don't know that I'm a huge fan of just checking the type and just handling different types completely differently. This proto shouldn't have any concept of how the EKF works, it should just hold information that the EKF uses, like encoder counts, if that helps answer your question?

File: src/shared/proto/tbots_frankie_status.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_frankie_status.proto]: Agreed

File: src/shared/proto/BUILD
Comment (MathewMacDougall) [reply src/shared/proto/BUILD]: Yea I was thinking of how "bad" it is for the EKF to work around the proto. Would a single EKF be able to support two message types just with a different interface, or would we need 1 EKF per type? Agree that our proto design shouldn't make assumptions about how it will be used in filtering.

File: src/shared/proto/tbots_frankie_status.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_frankie_status.proto]: @jonathanlew mind adding that to the style guide docs in this PR then?

File: src/shared/proto/tbots_robot_status_2014.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_robot_status_2014.proto]: The only field I've seen used in any way is the `fw_build_id`, where we would get warnings if the ID mismatched with running software. Everything else I've never seen used (in any real capacity). And in the last 2-3 years I haven't even seen that error, so we either changed something to no longer use it (maybe broke it porting the radio/annunciator during the ROS rewrite), or we haven't had mismatched builds in a long time.

Unless we plan on supporting a similar build ID system going forwards (@garethellis0 @akhilveeraghanta ), I'd say remove it. It's not useful in the legacy stack, and if we won't use it in the future no point having it.

File: src/shared/proto/tbots_frankie_status.proto
Comment (jonathanlew) [reply src/shared/proto/tbots_frankie_status.proto]: done

File: src/shared/proto/BUILD
Comment (garethellis0) [reply src/shared/proto/BUILD]: I'm a bit confused about what you're asking here. The EKF shouldn't be related at all to the proto, both robots will supply encoder readings in the end, so there's no difference in the information that makes it to the EKF.

File: src/shared/proto/tbots_robot_status_2014.proto
Comment (garethellis0) [reply src/shared/proto/tbots_robot_status_2014.proto]: The build ID system is a very good idea IMO, especially if we ever move to sharing dynamic params between the computer and the robot, we definitely want to know if we have a mismatch.

File: src/shared/proto/BUILD
Comment (MathewMacDougall) [reply src/shared/proto/BUILD]: I'm also confused, lets discuss in the meeting today.

**Comment by jonathanlew** [comment]
@garethellis0 @MathewMacDougall @akhilveeraghanta I've updated the structure of the protos, please re-review

File: src/shared/proto/tbots_robot_msg.proto
Comment (MathewMacDougall) [inline src/shared/proto/tbots_robot_msg.proto]: There's `EncoderData` here at the top level as well as within `MotorStatus`, which I think is confusing. I think my preference is to just have it at the top level, so it's not coupled to the `MotorStatus`

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Feels kind of odd that `MotorStatus` only has fields for the dribbler (ignoring encoder stuff). Not sure if I'd rather split dribbler stuff into it's own message or not.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Overall structure seems reasonable, main thing is just some slightly odd grouping IMO.

Overall I do like how you've created messages more by "concept" like temperature, because I think that's more agnostic of hardware and therefore more scalable / compatible.

However like I mentioned in one of my comments, it (currently) feels weird to have `MotorStatus` only contain dribbler speed info. Then we also have some dribbler info in `Temperature`. I'm not sure if grouping this info into a `Dribbler` message is the way to go or not, because then you lose consistency with the rest of the messages.

Definitely would like some other opinions on this.

**Comment by garethellis0** [comment]
Before I review, are we we looking at removing stuff we don't need anymore as well here? There seems to be a ton of extra stuff we can get rid of.

**Comment by jonathanlew** [comment]
> Before I review, are we we looking at removing stuff we don't need anymore as well here? There seems to be a ton of extra stuff we can get rid of.

Yup. I just included everything here for completeness and to make sure we're ok with removing things. The RadioBackend will ignore the fields that exist in the radio packet but are not part of the TbotsRobotMsg proto

**Comment by MathewMacDougall** [comment]
@jonathanlew do you want us to point out stuff we don't think we need in these next reviews then? Personally I think that's just worth another call for speed.

**Comment by jonathanlew** [comment]
> @jonathanlew do you want us to point out stuff we don't think we need in these next reviews then? Personally I think that's just worth another call for speed.

Call makes sense

File: src/shared/proto/tbots_robot_msg.proto
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_robot_msg.proto:5]: Should we add a test that builds this message with all the fields required for the old robots and verifies that it doesn't go over 100 bytes like Gareth suggested? I feel like we are getting pretty close if not already over the limit

**Review by akhilveeraghanta — COMMENTED** [review]
Overall I really like the new structure, nice work :fire:!

We have a lot of required fields, this may or may not be an issue, but since we are following googles style guide, just wanted to highlight this comment they had on required vs optional

"You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field – old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally. You should consider writing application-specific custom validation routines for your buffers instead. Some engineers at Google have come to the conclusion that using required does more harm than good; they prefer to use only optional and repeated. However, this view is not universal."

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Not sure what the range of this field is atm but IMO we should normalize it to [0, 1] and then we can remove the `break_beam_scale` since I don't think it's useful, but this may be a question for ELEC too.

I also think this field should be a `float`, since there's no need for `double` precision

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto:19]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto:23]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it. I don't see us ever in the case where the robot's max kick speed is less than the rules allow.

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it. At best this should be a constant somewhere else.

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it. We're not changing the FPGA (and won't have one in the new gen)

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto:87]: Not sure if this is useful in tandem with the `fw_build_id`. @garethellis0 thoughts?

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Not sure this is useful, depending on what "power level" means here. Assuming it's rpm, I don't think this is useful, and that this should probably just be a constant.

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: nit: add units to name -> `dribbler_rpm`

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful and we can remove it

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: I don't think this field is useful. It should be pretty obvious if the robot is alive if the AI is receiving status messages at all, I don't think an extra bool is useful.

File: src/shared/proto/robot_component_status.proto
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: @akhilveeraghanta I'm assuming this field is useful for when we do filtering to get a timestep / timestamp for new values like encoders? Just checking because otherwise not sure how useful this is.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Ok left some comments for fields I don't think are needed in the proto

File: src/shared/proto/robot_component_status.proto
Comment (akhilveeraghanta) [reply src/shared/proto/robot_component_status.proto]: When Jon and I were discussing how we detect timeouts and get better metrics on latency, we thought including the time_created in the packet (because lwip supports sntp) is a really good diagnostic value to get that sort of information. The time would be synced up between the STM32 and the AI because they are on the same network.

We could also use it for filtering, but its mainly just for diagnostics

File: src/shared/proto/robot_component_status.proto
Comment (akhilveeraghanta) [reply src/shared/proto/robot_component_status.proto]: Just as a note, we can get a similar value for WiFi which tells us how far away from the router the robot is. But I agree we probably don't have a use for it

File: src/shared/proto/tbots_robot_msg.proto
Comment (MathewMacDougall) [reply src/shared/proto/tbots_robot_msg.proto:5]: testing replying to a comment @garethellis0

File: src/shared/proto/BUILD
Comment (MathewMacDougall) [reply src/shared/proto/BUILD]: wow look at me replying to comments @garethellis0

**Comment by jonathanlew** [comment]
Will reopen so Gareth can comment

**Pull request closed without merging**
